File: frontend/src/api/oauthConsent.ts
/**
 * Task M3c-1 — frontend API client for the SPA-driven OAuth consent flow.
 *
 * The MCP authorization server (jvspatial M3a) 302-redirects an
 * unauthenticated authorize request to the SPA route
 * `/oauth/authorize?<oauth params>`. The authed consent page reads those
 * params and talks to the M3b-1 backend consent endpoints:
 *
 *   - GET  /api/oauth/consent          → ConsentDetails (client + scopes)
 *   - POST /api/oauth/consent/approve  → { redirect } (URL to send the
 *                                         browser back to the MCP client)
 *
 * The shared axios client (`./client`) prefixes `/api` and attaches the JWT
 * bearer, so both calls are authenticated as the logged-in user.
 */
import api from './client';

/** Shape returned by GET /oauth/consent (backend M3b-1). */
export interface ConsentDetails {
  client_id: string;
  client_name: string;
  scopes: string[];
  redirect_uri: string;
  state: string;
}

/** URL the browser is sent to after an approve/deny decision. */
export interface ConsentRedirect {
  redirect: string;
}

/** OAuth params forwarded from the authorize redirect's query string. */
export type OAuthParams = URLSearchParams | Record<string, string>;

function toRecord(params: OAuthParams): Record<string, string> {
  if (params instanceof URLSearchParams) {
    return Object.fromEntries(params.entries());
  }
  return params;
}

/**
 * Fetch the consent screen details for the pending authorize request. The
 * oauth params (client_id, redirect_uri, scope, state, code_challenge, …)
 * are forwarded verbatim as the query string so the backend can validate
 * the registered client + redirect_uri before disclosing the client name.
 */
export async function getConsent(params: OAuthParams): Promise<ConsentDetails> {
  const { data } = await api.get<ConsentDetails>('/oauth/consent', {
    params: toRecord(params),
  });
  return data;
}

/**
 * Record an approval. The backend mints an authorization code and returns
 * the `redirect_uri?code=...&state=...` URL to send the browser to.
 */
export async function approve(params: OAuthParams): Promise<ConsentRedirect> {
  const { data } = await api.post<ConsentRedirect>('/oauth/consent/approve', {
    decision: 'approve',
    ...toRecord(params),
  });
  return data;
}

/**
 * Record a denial. The backend returns the
 * `redirect_uri?error=access_denied&state=...` URL.
 */
export async function deny(params: OAuthParams): Promise<ConsentRedirect> {
  const { data } = await api.post<ConsentRedirect>('/oauth/consent/approve', {
    decision: 'deny',
    ...toRecord(params),
  });
  return data;
}

File: backend/app/schemas/oauth_consent.py
"""Request/response shapes for the SPA-driven OAuth consent endpoints (M3b-1).

These models back the two bearer-authed endpoints in
``app/api/oauth_consent.py`` that let the Integral SPA drive the OAuth
consent step (fetch consent details + approve/deny). They mirror the OAuth
parameters jvspatial's authorization server validates on the authorize request
so the SPA can re-submit the exact request the consent-details fetch validated.

Per AGENTS.md § jvspatial Object-Spatial Contract, request/response shapes live
here, never inline in the handler module.
"""

from __future__ import annotations

from typing import List

from pydantic import BaseModel, Field


class ConsentDetailsResponse(BaseModel):
    """Validated consent details rendered by the SPA's consent screen.

    Returned by ``GET /oauth/consent``. The ``scopes`` are the client-filtered
    scope the authorization server validated (``grant.request.scope``), not the
    raw query value, so the SPA shows exactly what the AS would grant.
    """

    client_id: str = Field(..., description="The requesting client's id")
    client_name: str = Field(
        default="", description="Human-readable client name for the consent UI"
    )
    scopes: List[str] = Field(
        default_factory=list,
        description="Client-filtered scopes the AS validated for this request",
    )
    redirect_uri: str = Field(
        ..., description="The validated redirect URI the code will be returned to"
    )
    state: str = Field(
        default="", description="Opaque client state echoed back on redirect"
    )


class ApproveConsentRequest(BaseModel):
    """The consent decision plus the OAuth parameters to re-submit to the AS.

    Posted to ``POST /oauth/consent/approve``. The parameters re-carry the exact
    authorization request the consent-details fetch validated. ``permissions`` /
    ``scope`` here are NEVER trusted for permission computation — the granted
    permissions come solely from the server-resolved session user (see the
    handler's TRUST BOUNDARY note). ``scope`` is only the client-requested scope
    the AS intersects with the resolved permissions.
    """

    decision: str = Field(
        ...,
        description="'approve' issues a code; anything else denies",
    )
    client_id: str = Field(..., description="The requesting client's id")
    redirect_uri: str = Field(..., description="The redirect URI from the request")
    scope: str = Field(default="", description="Space-delimited requested scope")
    state: str = Field(default="", description="Opaque client state")
    code_challenge: str = Field(default="", description="PKCE code challenge (S256)")
    code_challenge_method: str = Field(
        default="S256", description="PKCE method (S256 only)"
    )
    response_type: str = Field(default="code", description="OAuth response type")


class ApproveConsentResponse(BaseModel):
    """The redirect URL the SPA should send the browser to.

    Returned by ``POST /oauth/consent/approve``. On approve it carries the
    issued ``code`` (+ ``state``); on deny it carries ``error=access_denied``.
    """

    redirect: str = Field(
        ..., description="redirect_uri with code+state, or error=access_denied"
    )


__all__ = [
    "ApproveConsentRequest",
    "ApproveConsentResponse",
    "ConsentDetailsResponse",
]
